Feat(#2) : Add Equalizer on axis renderer and add hands-off detector

manoukianv opened the pull request.

manoukianv committed:
docs: pure code cleanup, typing and Doxygen documentation

manoukianv committed:
feat: add torque slew rate limiter and driver calibration

manoukianv committed:
feat: refactor torque calculation into a 7-step pipeline

manoukianv committed:
feat: add ARM CMSIS-DSP library infrastructure and optimized filters

manoukianv committed:
feat: add FFB reconstruction filter and DSP optimizations

manoukianv committed:
feat: final TMC optimizations, project configs and build tools

manoukianv committed:
feat: migrate to new CMIS-DSP source code

manoukianv committed:
fix: git clone submodule for cmis-dsp

manoukianv committed:
feat: add equalizer and handsoff detection

manoukianv committed:
fix: add missing source in DSP for handsup

manoukianv committed:
fix: bug in slew rate when rotation is inverted.

manoukianv committed:
feat: add equalizer and handsoff detection

manoukianv committed:
fix: add missing source in DSP for handsup

manoukianv committed:
fix: fix the denormalisation issue in calculateFFBTorque. Optimize the struct from 64 to 32b to reduce load.

manoukianv committed:
feat(equalizer): change first and last filter type to improve the feedback.

Ultrawipf commented on Firmware/FFBoard/Src/Axis.cpp:
```diff
@@ -671,28 +781,39 @@ uint16_t Axis::getPower(){
-		return -powf(-torquef,expo) * 0x7fff;
-	}else{
-		return powf(torquef,expo) * 0x7fff;
+int32_t Axis::calculateFFBTorque(){
```
Converting back and forth between float and int32/16b int may lose resolution

Ultrawipf commented on Firmware/FFBoard/Src/Axis.cpp:
```diff
@@ -86,39 +101,62 @@ const std::vector<class_entry<MotorDriver>> Axis::axis2_drivers =
+		BiquadType type = BiquadType::peak;
+		if (idx == 0) type = BiquadType::lowshelf;
+		else if (idx == num_eq_bands - 1) type = BiquadType::highshelf;
+		eqFilters[idx].setBiquad(type, eq_frequencies[idx] / filter_f, 2.14f, 0.0);
```
Eq filters are not recalculated when the sample rate changes by updateSamplerate.

Ultrawipf commented on Firmware/FFBoard/Src/Axis.cpp:
```diff
@@ -723,73 +845,132 @@ void Axis::setEffectTorque(int32_t torque) {
+	
+	// STEP 3: Add endstops
+	// Note: Historically endstops are added before the Master Scaler.
+	// If hard endstops (like Simucube) are desired, this should be moved to STEP 5.
```
it should absolutely not be moved after the scaling. everything MUST be scaled or you blow up your psu and bypass the scaler which in the case of the tmc is the actual current limit

manoukianv committed:
refactor: simplify Biquad filter implementations and TMC init

manoukianv committed:
fix: remove leaked anticogging declarations

manoukianv committed:
refactor(axis): unify DSP and non-DSP logic in applySpeedLimiterTorque

manoukianv committed:
fix: address PR review comments on EffectsCalculator

- Change forces array from int64_t to int32_t to avoid unnecessary 64-bit math overhead.

- Deduplicate sine phase calculation using MATH_PI and MATH_SIN macros.

- Restore magnitude/offset assignment in HidFFB instead of EffectsCalculator to maintain separation of concerns.

manoukianv committed:
fix: use fallbackValue when reconstruction spline is not ready

- Prevents breaking SerialFFB and other sources that don't push samples to the reconstruction filter.

manoukianv committed:
refactor: deduplicate DSP math logic in HidFFB

- Use MATH_PI, MATH_SIN, and MATH_COS macros to remove duplicate ifdef blocks for direction calculations.

- Fix Z-axis calculation missing DSP optimizations.

manoukianv committed:
docs: update author list in FFBoardMain.cpp

manoukianv committed:
refactor: remove unused getVelocityControllerTorque function

manoukianv committed:
perf: skip magnitude interpolation and envelope calculation for Ramp effects

manoukianv committed:
build: remove CMSIS-DSP virtual pin from F407VG IOC

manoukianv committed:
refactor: optimize CMSIS-DSP header inclusions

manoukianv committed:
fix(axis): add missing statistics_functions.h for arm_std_f32

manoukianv committed:
fix: fix bug injected with the 64_t structure, 32b is enough with less cpu overload.
